AEXS/aex-141.md

```diff
@@ -300,124 +314,111 @@ ApprovalForAll(address, address, string)
 | operator | address |
 | approved | string |
 
-## Extension Mintable ("mintable")
+# Receiver contract interface
 
-### mint\(\)
+The standard only allows safe transfers of tokens. On transfer a check MUST be performed which checks if the recipient is a contract and if so the transfer MAY ONLY happen if `on_aex141_received` returns true.
 
-Issues a new token to the provided address. If the `owner` is a contract, NFTReceiver will be called with `data` if provided.
-Emits a Transfer event.
-Throws if the call to NFTReceiver implementation failed (safe transfer). 
+## AEX141Receiver
 
 ```sophia
-stateful entrypoint mint(owner: address, metadata: option(metadata), data: option(string)) : int
+contract interface IAEX141Receiver = 
+    entrypoint on_aex141_received : (option(address), int, option(string)) => bool
 ```
 
-| parameter | type | 
-| :--- | :--- | 
-| owner | address |
-| metadata | option(metadata) |
-| data  | option(string) | 
+### on_aex141_received\(\)
 
-| return | type |
-| :--- | :--- |
-| token_id | int |
-
-## Extension Burnable ("burnable")
-
-### burn\(\)
+Deals with receiving NFTs on behalf of a contract. Contracts MUST implement this interface to be able to receive NFTs. Mint and transfer transactions will invoke the `on_aex141_received` function.
 
-Burns the NFT with the provided `token_id`.
-Emits the `Transfer` event.
+Returns `true` or `false` to signal whether processing the received NFT was successful or not.
 
 ```sophia
-stateful entrypoint burn(token_id: int) : unit
+entrypoint on_aex141_received(from: option(address), token_id: int, data: option(string)) : bool
 ```
-
 | parameter | type |
 | :--- | :--- |
+| from | option(address) |
 | token_id | int |
+| data | option(string) |
 
-## Extension Swappable ("swappable")
+# Extensions
 
-### swap\(\)
```

Comment:
with the removal of swap we don't expect the reference tokens to be able to be swapped any more?

Reply:
sorry, missed to answer that question.

well, we just don't define any standardized way of doing it. we can reconsider this if we finally can provide good and clean examples using the new polymorphism feature of Sophia.

people could still include that in a custom implementation but we do not expect the `Swap` event in any way right now.

as you pointed out all burns (wherever they might happen) should emit `Burn` though

Reply:
> yeah, swapping will then be more complicated, but ok for now

I still have the implementation in the history of the example repo. we can reconsider this as soon as we have examples making use of polymorphism to implement interfaces of extensions.

as we decided we do not want to advertise usage of `Swappable` I assumed we agreed on removing it completely - at least for now